lib/Runtime/Library/JavascriptRegExpConstructor.cpp

```diff
@@ -53,17 +54,44 @@ namespace Js
         this->lastInput = lastInput;
         this->lastMatch = lastMatch;
         this->reset = true;
+        this->invalidatedLastMatch = false;
+    }
+
+    void JavascriptRegExpConstructor::InvalidateLastMatch(UnifiedRegex::RegexPattern* lastPattern, JavascriptString* lastInput)
+    {
+        AssertMsg(lastPattern != nullptr, "lastPattern should not be null");
+        AssertMsg(lastInput != nullptr, "lastInput should not be null");
+        AssertMsg(JavascriptOperators::GetTypeId(lastInput) != TypeIds_Null, "lastInput should not be JavaScript null");
+
+        this->lastPattern = lastPattern;
+        this->lastInput = lastInput;
+        this->lastMatch.Reset();
+        this->reset = true;
+        this->invalidatedLastMatch = true;
     }
 
     void JavascriptRegExpConstructor::EnsureValues()
     {
         if (reset)
         {
-            Assert(!lastMatch.IsUndefined());
             ScriptContext* scriptContext = this->GetScriptContext();
+            const CharCount lastInputLen = lastInput->GetLength();
+            const char16* lastInputStr = lastInput->GetString();
             UnifiedRegex::RegexPattern* pattern = lastPattern;
+
+            // When we perform a regex test operation it's possible the result of the operation will be loaded from a cache and the match will not be computed and updated in the ctor.
+            // In that case we invalidate the last match stored in the ctor and will need to compute it before it will be accessible via $1 etc.
+            // Since we only do this for the case of RegExp.prototype.test cache hit, we know several things:
+            //  - The regex is not global or sticky
+            //  - There was a match (test returned true)
+            if (invalidatedLastMatch)
+            {
+                this->lastMatch = RegexHelper::SimpleMatch(scriptContext, pattern, lastInputStr, lastInputLen, 0);
```

Comment:
>SimpleMatch [](start = 47, length = 11)

could this function be re-entrant and change lastInput by any chance?

Reply:
I hope not. We don't pass any `Var` here, just the `const char16*` so I don't think user can apply getter/setters or use Symbol overrides. Seems as safe as the other operations here.